scheduling/allocator.rb

```diff
@@ -0,0 +1,253 @@
+        .where { (total_cores - used_cores >= request.cores) }
+        .with(:total_ipv4, DB[:address]
+          .select_group(:routed_to_host_id)
+          .select_append { round(sum(power(2, 32 - masklen(cidr)))).cast(:integer).as(total_ipv4) }
+          .where { (family(cidr) =~ 4) })
+        .with(:used_ipv4, DB[:address].left_join(:assigned_vm_address, address_id: :id)
+          .select_group(:routed_to_host_id)
+          .select_append { (count(Sequel[:assigned_vm_address][:id]) + 1).as(used_ipv4) })
+        .with(:storage_devices, DB[:storage_device]
+          .select_group(:vm_host_id)
+          .select_append { count.function.*.as(num_storage_devices) }
+          .select_append { sum(available_storage_gib).as(available_storage_gib) }
+          .select_append { sum(total_storage_gib).as(total_storage_gib) }
+          .select_append { json_agg(json_build_object(Sequel.lit("'id'"), Sequel[:storage_device][:id], Sequel.lit("'total_storage_gib'"), total_storage_gib, Sequel.lit("'available_storage_gib'"), available_storage_gib)).order(available_storage_gib).as(storage_devices) }
+          .where(enabled: true)
+          .having { sum(available_storage_gib) >= request.storage_gib }
+          .having { count.function.* >= (request.distinct_storage_devices ? request.storage_volumes.count : 1) })
+
+      ds = ds.where { used_ipv4 < total_ipv4 } if request.ip4_enabled
+      ds = ds.where(vm_host_id: request.host_filter) unless request.host_filter.empty?
+      ds = ds.where(location: request.location_filter) unless request.location_filter.empty?
+      ds.all
+    end
+
+    def self.update_vm(vm_host, vm)
+      ip4, address = vm_host.ip4_random_vm_network if vm.ip4_enabled
+      fail "no ip4 addresses left" if vm.ip4_enabled && !ip4
+      vm.update(
+        vm_host_id: vm_host.id,
+        ephemeral_net6: vm_host.ip6_random_vm_network.to_s,
+        local_vetho_ip: vm_host.veth_pair_random_ip4_addr.to_s,
+        allocated_at: Time.now
+      )
+      AssignedVmAddress.create_with_id(dst_vm_id: vm.id, ip: ip4.to_s, address_id: address.id) if ip4
+      vm.sshable&.update(host: vm.ephemeral_net4 || vm.ephemeral_net6.nth(2))
+    end
+
+    def initialize(host, request)
+      @host = host
+      @request = request
+      @allocations = [
+        HostAllocation.new(:used_cores, host[:total_cores], host[:used_cores], request.cores),
+        HostAllocation.new(:used_hugepages_1g, host[:total_hugepages_1g], host[:used_hugepages_1g], request.mem_gib),
+        StorageAllocation.new(host, request)
+      ]
+      @score = calculate_score
+    end
+
+    def is_valid
+      @allocations.all? { _1.is_valid }
+    end
+
+    def update(vm)
+      vm_host = VmHost[@host[:vm_host_id]]
+      DB.transaction do
+        Allocation.update_vm(vm_host, vm)
+        VmHost.dataset.where(id: @host[:vm_host_id]).update(@allocations.map { _1.update(vm, vm_host) }.reduce(&:merge))
```

Comment:
I misunderstood `@allocations.map { _1.update(vm, vm_host) }.reduce(&:merge)` initially. Just by looking at how `allocation::update` is used, I thought it returns columns in the VmHost to be updated, which was not accurate since it also has some side-effects in `StorageAllocation::update`.

What do you think breaking `allocation::update` into two methods? Like:
* `update` (which creates necessary database records)
* `vm_host_changes`

I think this way it might be a bit easier to understand without looking at the definitions of methods being called.

Reply:
split this up into two different methods